### PR TITLE
[#17] fix sites that require headers and auth

### LINKED ISSUE #17: Cypress web vitals will not work if credentials or special headers needed to visit the site.
# Issue
Cypress web vitals will not work if credentials or special headers needed to visit the site.
```javascript
cy.visit({
      url: 'https://www.someQaEnvironment.com/',
      headers: {
        "Accept-Encoding": "gzip, deflate"
      },
      auth: {
        username: 'qa',
        password: 'examplepass',  
      }
 });
```

If these are not provided, `ESocket Timeout error`
Took a stab at a fix that worked for me locally so pr coming soon! 🎉
currently in a draft pr: #18 

Setup:

- node version: v16.14.2
- cypress-web-vitals version: 2.0.0
 "cypress-real-events": "^1.7.1",

## Details
1) visitWithWebVitalsSnippet, runs cy.visit after intercept but loses auth and header context.
2) cy.intercept is not grabbing auth param.
3) cy.intercept gives different headers than are required by cy.visit
```javascript
// cy.visit needs:
      headers: {
        "Accept-Encoding": "gzip, deflate"
      },
```

```javascript
// cy.intercept gets:
      headers: {
        accept: "text/html,*/*"
        authorization: "Basic cWF1c2VyOlRlc3RlZDRxdWFsaXR5IQ=="
        connection: "keep-alive"
        cookie:  "some cookie"
        host: "www.example.com"
        user-agent: "some agent"
      }
```
^ because of ( 3) ) , cy.vitals is easy to fix 
but cy.visit is not an easy fix so my draft pr #18  only fixes cy.vitals

### PR DESCRIPTION
**Pull request by MichaelDimmitt**
# Issue

Fixes #17 

## Details

Allow core web vitals to work on websites that require auth and headers.

#### CY.vitals will work with params passed
```javascript
cy.vitals({
      url: 'https://www.someQaEnvironment.com/',
      headers: {
        "Accept-Encoding": "gzip, deflate"
      },
      auth: {
        username: 'qa',
        password: 'examplepass',  
      }
 }); ✅
```

#### Note: CY.visit will still not work:
```javascript
cy.visit({
      url: 'https://www.someQaEnvironment.com/',
      headers: {
        "Accept-Encoding": "gzip, deflate"
      },
      auth: {
        username: 'qa',
        password: 'examplepass',  
      }
 }); 🚫
```
#### TODO:
1) update tests
2) update docs to introduce new options on `cy.vitals({})`

## CheckList

- [x] PR starts with [#_ISSUE_ID_].
- [x] Has been tested (where required).

### CODE REVIEW

**Comment by cmorten** [comment]
@MichaelDimmitt is this ready for review? Clearing notifications and noticed this is still in draft?

**Comment by MichaelDimmitt** [comment]
It works as is and feel free to give feedback

I will try to finish it today so that tests are passing.
And will double check to see if there are any breaking changes for people doing custom things.

1) fix tests
2) update docs
3) make sure I broke no custom people.

**Comment by MichaelDimmitt** [comment]
```javascript
const visitWithWebVitalsSnippet = (url, auth = undefined, headers = undefined) => {
  cy.intercept({ method: "GET", url, times: 1 }, (req) => {
    req.continue((res) => {
      const body = res.body.replace(
        /<head(>|.*?[^?]>)/,
        `<head$1${WEB_VITALS_SNIPPET}`
      );
      res.send(body);
    });
  });
  if(auth && headers) {
    return cy.visit(url, { auth, headers, log: false });  
  }
  else {
    return cy.visit(url, { log: false });
  }
};
```
debating between these two options:
```javascript
const visitWithWebVitalsSnippet = (url, auth = null, headers = {}) => {
  cy.intercept({ method: "GET", url, times: 1 }, (req) => {
    req.continue((res) => {
      const body = res.body.replace(
        /<head(>|.*?[^?]>)/,
        `<head$1${WEB_VITALS_SNIPPET}`
      );
      res.send(body);
    });
  });

  return cy.visit(url, { auth, headers, log: false });
};
```

**Comment by MichaelDimmitt** [comment]
@cmorten , can you trigger a test re-run

File: src/visitWithWebVitalsSnippet.js [hunk 1]
```diff
@@ -10,8 +10,7 @@ const visitWithWebVitalsSnippet = (url) => {
       res.send(body);
     });
   });
-
-  return cy.visit(url, { log: false });
+  return cy.visit(url, { ...rest, log: false });
```
Comment (cmorten) [inline src/visitWithWebVitalsSnippet.js:13]:
```suggestion

  return cy.visit(url, { ...rest, log: false });
```

File: src/visitWithWebVitalsSnippet.test.js [hunk 2]
```diff
@@ -12,84 +12,120 @@ describe("visitWithWebVitalsSnippet", () => {
       visit: jest.fn(),
     };
 
-    visitWithWebVitalsSnippet(mockUrl);
+    visitWithWebVitalsSnippet({ url: mockUrl });
   });
+  describe("only a url was passed", () => {
```
Comment (cmorten) [inline src/visitWithWebVitalsSnippet.test.js:17]:
```suggestion
  });

  describe("only a url was passed", () => {
```

File: src/visitWithWebVitalsSnippet.test.js [hunk 3]
```diff
@@ -12,84 +12,121 @@ describe("visitWithWebVitalsSnippet", () => {
...
-        `<html><head>${WEB_VITALS_SNIPPET}<title>App</title><script>/** comment with <head> in it **/</script></head><body></body></html>`
-      );
+    it("should visit the url without logging", () => {
+      expect(global.cy.visit).toHaveBeenCalledWith(mockUrl, { log: false });
     });
   });
-
-  it("should visit the url without logging", () => {
-    expect(global.cy.visit).toHaveBeenCalledWith(mockUrl, { log: false });
+  describe("additional params passed like auth + headers", () => {
```
Comment (cmorten) [inline src/visitWithWebVitalsSnippet.test.js:97]:
```suggestion

  describe("additional params passed like auth + headers", () => {
```

File: src/visitWithWebVitalsSnippet.test.js [hunk 4]
```diff
@@ -12,84 +12,122 @@ describe("visitWithWebVitalsSnippet", () => {
...
+          auth: examplAuth,
+          headers: exampleHeaders,
+        });
+      });
+      it("should visit the url without logging", () => {
+        expect(global.cy.visit).toHaveBeenCalledWith(mockUrl, {
+          headers: {
+            "Accept-Encoding": "gzip, deflate",
+          },
+          auth: {
```
Comment (cmorten) [inline src/visitWithWebVitalsSnippet.test.js:124]:
```suggestion
      const exampleHeaders = {
        "Accept-Encoding": "gzip, deflate",
      };
      const exampleAuth = {
        username: "test-username",
        password: "test-password",
      };

      beforeEach(() => {
        jest.clearAllMocks();

        global.cy = {
          intercept: jest.fn(),
          visit: jest.fn(),
        };

        visitWithWebVitalsSnippet({
          url: mockUrl,
          auth: exampleAuth,
          headers: exampleHeaders,
        });
      });

      it("should visit the url without logging", () => {
        expect(global.cy.visit).toHaveBeenCalledWith(mockUrl, {
          headers: {
            "Accept-Encoding": "gzip, deflate",
          },
          auth: {
```

File: src/visitWithWebVitalsSnippet.test.js [hunk 5]
```diff
@@ -12,84 +12,122 @@ describe("visitWithWebVitalsSnippet", () => {
...
+            username: "qa",
+            password: "examplepass",
```
Comment (cmorten) [inline src/visitWithWebVitalsSnippet.test.js:126]:
```suggestion
            username: exampleAuth.username,
            password: exampleAuth.password,
```

File: src/visitWithWebVitalsSnippet.test.js [hunk 6]
```diff
@@ -12,84 +12,122 @@ describe("visitWithWebVitalsSnippet", () => {
...
+          },
```
Comment (cmorten) [inline src/visitWithWebVitalsSnippet.test.js:127]:
```suggestion
      const mockHeaders = {
        "Accept-Encoding": "gzip, deflate",
      };
      const mockAuth = {
        username: "qa",
        password: "examplepass",
      };

      beforeEach(() => {
        jest.clearAllMocks();

        global.cy = {
          intercept: jest.fn(),
          visit: jest.fn(),
        };

        visitWithWebVitalsSnippet({
          url: mockUrl,
          auth: examplAuth,
          headers: exampleHeaders,
        });
      });

      it("should visit the url without logging", () => {
        expect(global.cy.visit).toHaveBeenCalledWith(mockUrl, {
          headers: mockHeaders,
          auth: mockAuth,
```

File: src/visitWithWebVitalsSnippet.test.js [hunk 7]
```diff
@@ -12,84 +12,121 @@ describe("visitWithWebVitalsSnippet", () => {
...
+
+        global.cy = {
+          intercept: jest.fn(),
+          visit: jest.fn(),
+        };
+
+        visitWithWebVitalsSnippet({
+          url: mockUrl,
+          auth: examplAuth,
+          headers: exampleHeaders,
```
Comment (cmorten) [inline src/visitWithWebVitalsSnippet.test.js:119]:
```suggestion
          auth: mockAuth,
          headers: mockHeaders,
```

**Comment by MichaelDimmitt** [comment]
🚀

**Pull request merged**